### ISSUE TITLE
[clang asan] AddressSanitizer fails to report buffer over-read in sscanf on an unterminated buffer

### ISSUE THREAD

**Issue by tkuchta** [llvm/llvm-project#53635]
Hi, 

I have spotted a potential asan issue: there is a stack char buffer ```buf``` which is filled with a value 0x1 and is not null-terminated [1, 2]. 
We scan the buffer with ```sscanf``` searching for an integer value [3]. Since the buffer contains only 0x1s, the search is going to go out of bounds. However, asan doesn't detect the issue.
I have verified the problem on a recent clang build (version 14).

```
#include <stdio.h>
#include <string.h>

int main(int argc, char* argv[]) {
        char buf[20] =  { 0 };                  // [1]
        memset(buf, 1, sizeof(buf));        // [2]
        int outval = 0;
        int ret = sscanf(buf, "%d", &outval); // [3]
        printf("ret is %d, outval is %d\n", ret, outval);
        return outval;
}
```

Compilation:
```
clang -fsanitize=address test.c -o test
```

Interestingly, ```valgrind``` detects an issue when buffer size is 20 but doesn't detect it for size 10:
```
$ valgrind ./test
==1852734== Memcheck, a memory error detector
==1852734== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==1852734== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==1852734== Command: ./test
==1852734==
==1852734== Conditional jump or move depends on uninitialised value(s)
==1852734==    at 0x4843596: rawmemchr (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1852734==    by 0x48F7F35: _IO_str_init_static_internal (strops.c:41)
==1852734==    by 0x48C74C2: _IO_strfile_read (strfile.h:95)
==1852734==    by 0x48C74C2: __isoc99_sscanf (isoc99_sscanf.c:28)
==1852734==    by 0x4011A8: main
```
